lib/spack/spack/util/elf.py

```diff
@@ -0,0 +1,455 @@
+            elf.rpath_strtab_offset = val
+            elf.dt_rpath_offset = current_offset
+            elf.is_runpath = True
+            elf.has_rpath = True
+        elif tag == ELF_CONSTANTS.DT_STRTAB:
+            count_strtab += 1
+            strtab_vaddr = val
+        elif tag == ELF_CONSTANTS.DT_NEEDED:
+            elf.has_needed = True
+            elf.dt_needed_strtab_offsets.append(val)
+        elif tag == ELF_CONSTANTS.DT_SONAME:
+            elf.has_soname = True
+            elf.dt_soname_strtab_offset = val
+        current_offset += dynamic_array_size
+
+    # No rpath/runpath, that happens.
+    if count_rpath == count_runpath == 0:
+        elf.has_rpath = False
+    elif count_rpath + count_runpath != 1:
+        raise ElfParsingError("Could not find a unique rpath/runpath.")
+
+    if count_strtab != 1:
+        raise ElfParsingError("Could not find a unique strtab of for the dynamic section strings")
+
+    # Nothing to retrieve, so don't bother getting the string table.
+    if not (elf.has_rpath or elf.has_soname or elf.has_needed):
+        return
+
+    string_table = retrieve_strtab(f, elf, vaddr_to_offset(elf, strtab_vaddr))
+
+    if elf.has_needed:
+        elf.dt_needed_strs = list(
+            parse_c_string(string_table, offset) for offset in elf.dt_needed_strtab_offsets
+        )
+
+    if elf.has_soname:
+        elf.dt_soname_str = parse_c_string(string_table, elf.dt_soname_strtab_offset)
+
+    if elf.has_rpath:
+        elf.dt_rpath_str = parse_c_string(string_table, elf.rpath_strtab_offset)
+
+
+def parse_header(f, elf):
+    # Read the 32/64 bit class independent part of the header and validate
+    e_ident = f.read(16)
+
+    # Require ELF magic bytes.
+    if len(e_ident) != 16 or e_ident[:4] != ELF_CONSTANTS.MAGIC:
+        raise ElfParsingError("Not an ELF file")
+
+    # Defensively require a valid class and data.
+    e_ident_class, e_ident_data = get_byte_at(e_ident, 4), get_byte_at(e_ident, 5)
+
+    if e_ident_class not in (ELF_CONSTANTS.CLASS32, ELF_CONSTANTS.CLASS64):
+        raise ElfParsingError("Invalid class found")
+
+    if e_ident_data not in (ELF_CONSTANTS.DATA2LSB, ELF_CONSTANTS.DATA2MSB):
+        raise ElfParsingError("Invalid data type")
+
+    elf.is_64_bit = e_ident_class == ELF_CONSTANTS.CLASS64
+    elf.is_little_endian = e_ident_data == ELF_CONSTANTS.DATA2LSB
+
+    # Set up byte order and types for unpacking
+    elf.byte_order = "<" if elf.is_little_endian else ">"
+
+    # Parse the rest of the header
+    elf_header_fmt = elf.byte_order + ("HHLQQQLHHHHHH" if elf.is_64_bit else "HHLLLLLHHHHHH")
+    hdr_size = calcsize(elf_header_fmt)
+    data = read_exactly(f, hdr_size, "ELF header malformed")
+    elf.elf_hdr = ElfHeader._make(unpack(elf_header_fmt, data))
+
+
+def _do_parse_elf(f, interpreter=True, dynamic_section=True):
+    # We currently do not deal with ELF files with an offset in the file; offsets are
+    # always absolute.
+    f.seek(0)
```

Comment:
This should be harmless but confuses me slightly, the default for seek is to seek based on offset 0 rather than from current anyway, is there a specific issue you're trying to work around here?

Reply:
My idea was this: someone passes an `f` not at offset 0; they already ran read/seek. Then the parser reads the ELF header, and does absolute fseek based on what offets it finds in the header, which is wrong. Instead here we just require that the header itself starts at 0.

Reply:
Alternatively, I can also make it defensively throw if `f.tell() != 0`

Reply:
Got it, makes sense to me. 👍